packages/web3-eth-contract/src/contract.ts

```diff
@@ -427,127 +363,299 @@ export class Contract<Abi extends ContractAbi>
 			set: (value: ContractAbi) => this._parseAndSetJsonInterface(value, returnDataFormat),
 			get: () => this._jsonInterface,
 		});
+
+		this.setConfig(Contract.contractWeb3Config);
 	}
 
+	/**
+	 * Can be used to set {@link Contract.defaultAccount} for all contracts.
+	 */
+	public static get defaultAccount() {
+		return (this.constructor as typeof Contract).defaultAccount;
+	}
+
+	public static set defaultAccount(value: Address | undefined) {
+		(this.constructor as typeof Contract).defaultAccount = value;
+		if (value === undefined) delete Contract.contractWeb3Config.defaultAccount;
+		else Contract.contractWeb3Config.defaultAccount = value;
+	}
 	public get defaultAccount() {
-		return (this.constructor as typeof Contract).defaultAccount ?? super.defaultAccount;
+		return super.defaultAccount ?? (this.constructor as typeof Contract).defaultAccount;
 	}
 
 	public set defaultAccount(value: Address | undefined) {
 		super.defaultAccount = value;
```

Comment:
so currently we have defaults at two levels?
1. contracts shared level ( static )
2. global ( coming via inheritance context )

 we should have defaults as:

1. contract instance level ( copy when contract init from global to instance level first time) and always use these first in whole contract. ( change at contract instance level should **NOT** be reflected back to global level )

2. global level ( any change at global level should be reflected down to all contract instance levels )

Reply:
I believe 1 is fine in the implementation. 

Regarding 2. When an instance is initialized it gets the properties from the shared level 
https://github.com/web3/web3.js/pull/5756/files#diff-7fa8f782e68ce845dbb6e8446b86041697fa427db5485b9c63bf689d0b32ed81R367

However, if an instance is initialized and a static property is updated afterward it is not reflected down. I'm not sure how this could be done without having access to an `EventEmitter` ( I think it's not possible to use one in static methods).

Reply:
so in that case we can remove **static** from `contractWeb3Config` and just have:

1. instance level defaults ( no upward sync )
2. global level defaults  coming via config  ( with downward sync support in contracts , it will be very easy to create a flag in contract class and make it optional downward sync as new feature for end users)

and can sync down global level defaults just by subscribing `Web3ConfigEvent.CONFIG_CHANGE`  https://github.com/web3/web3.js/blob/498f07764a65e45fda2a902b8b7bd61aa70663c9/packages/web3-core/src/web3_config.ts#L519, contract can subscribe in its constructor to this event and sync its instance level defaults with that.

Reply:
```
//in constructor 
if( new_feature_flag)
    super.on(Web3ConfigEvent.CONFIG_CHANGE, event => {
          //sync instance level defaults with globals 
}
```

Reply:
> so in that case we can remove **static** from `contractWeb3Config` and just have:
> 
> 1. instance level defaults ( no upward sync )
So in the setter functions in `Contract` class, shouldn't use existing implementation
```ts
super.defaultValue = value
```
(Like in defaultHardfork here:https://github.com/web3/web3.js/pull/5756/files#diff-7fa8f782e68ce845dbb6e8446b86041697fa427db5485b9c63bf689d0b32ed81L453)

because it will use the setter of the `config` and it will trigger a `CONFIG_CHANGE` event which will be caught by all `context`/`config` instances.

> 2. global level defaults  coming via config  ( with downward sync support in contracts , it will be very easy to create a flag in contract class and make it optional downward sync as new feature for end users)
> 
> and can sync down global level defaults just by subscribing `Web3ConfigEvent.CONFIG_CHANGE`
> 
> https://github.com/web3/web3.js/blob/498f07764a65e45fda2a902b8b7bd61aa70663c9/packages/web3-core/src/web3_config.ts#L519
> 
> , contract can subscribe in its constructor to this event and sync its instance level defaults with that.

As it is now, we user `super.defaultValue`, so from what I get we are already subscribed.

I see three cases:

- contracts shared level ( static )
- contract instance level 
- global ( coming via inheritance context )

Now there is no difference between global and instance-level variables.
Do you think we should distinguish them?
At the instance level, how we should prioritize them (which value to use for sending transaction etc.) since we could have (two/ or three if we distinguish them)

Reply:
Thanks @Nikos for looking this in detail. 

yes there should be difference between global and instance level defaults. 
and first priority should be given to instance level in all cases during usage ( sending or preparing txs etc )

There are usecases for this, as a user can have multiple contracts with different defaults ( like default account etc )

Reply:
I propose to have the following:
The first place to have the default config (that could be read from .env file). This default configuration will have the following values:
```js
{
		handleRevert: false,
		defaultBlock: 'latest',
		transactionBlockTimeout: 50,
		transactionConfirmationBlocks: 24,
		transactionPollingInterval: 1000,
		transactionPollingTimeout: 750 * 1000,
		transactionSendTimeout: 750 * 1000,
		blockHeaderTimeout: 10,
		maxListenersWarningThreshold: 100,
		defaultChain: 'mainnet',
		defaultHardfork: 'london',
		defaultTransactionType: '0x0',
		defaultMaxPriorityFeePerGas: toHex(2500000000),
		enableExperimentalFeatures: {
			useSubscriptionWhenCheckingBlockTimeout: false,
			useRpcCallSpecification: false,
		}
}
```
The second place is the contract static values (all set to `undefined` by default).
The third place is the instance values (all set to `undefined` by default).


Any value of the previous 3 places could be changed. However, when reading the values the priorities are from the third place (instance) then the second (the static), then the first (the default values).
So, when the user would read a property's value, the code would check if the third place (at the instance level) had a value that is different from `undefined`, if yes it will return it. And if not, it will check the value at the second place (at the static properties). If the value is not `undefined` it will return it. If not, the code will reach to the value at the default configuration and return it.

This was there would be no need for listing to `CONFIG_CHANGE` event. And we can enable the user to ignore reading the instance level and use the static ones by a flag variable. Or probably it would be better if we add a function named `resetToDefault(toGlobalDefaults: boolean)` that reset the configurations either to the static values or the global default values based on the passed boolean variable (`toGlobalDefaults`).

Reply:
A bit irrelevant to the discussion, but I believe the following is a bug:

```typescript
const web3Context = new Web3Context();
web3Context.defaultBlock = 'pending'

const contract = new Contract([], web3Context);
console.log(contract.getConfig().defaultBlock);
// > latest
```

When `Contract` is passed an instance of `Web3Context`, [this config](https://github.com/web3/web3.js/blob/4bbc030ba86f2e5b6494b395d77acce130d90c66/packages/web3-core/src/web3_context.ts#L127) is `undefined` because the expected `Web3Context.config` is actually [_config](https://github.com/web3/web3.js/blob/4bbc030ba86f2e5b6494b395d77acce130d90c66/packages/web3-core/src/web3_config.ts#L68) and `Web3Config.getConfig` is expected to be used to access the `config` object

---

Discussed long ago, the reason why `Web3Config` uses `Web3EventEmitter` is so plugins and users could subscribe to value changes and ultimately decide how to handle them

Here are some of my thoughts/understandings:

1. `Web3Context` is the global control and source for defaults
    1.a The idea with this is that a user instantiates a single instance of `Web3Context` and can pass it to any `web3` package (that can make use of it) to configure it. If each package stores the instance of `Web3Context` as a reference, then all packages will get updates to default properties without having to do anything special
2. If `Contract` is instantiated with an instance of `Web3Context`, then it points to it as it's config and uses it's defaults (receiving updates as mentioned in `1.a`)
    2.a Else it defaults to the same config object `Web3Context` defaults to (configured by `Web3Config` [here](https://github.com/web3/web3.js/blob/4bbc030ba86f2e5b6494b395d77acce130d90c66/packages/web3-core/src/web3_config.ts#L68))
        2.a.1 This default object should probably be pulled into an exported object such as `Web3ConfigOptionsDefaults` from `web3-core/src/web3_config.ts`
3. We remove the `static` properties from `Contract`
4. Replace `Web3Config`'s `getConfig` and `setConfig` with a getter and setter
    4.a This would allow the config object to be accessed like `contract.config` instead of `contract.getConfig()` 

With changes `3` and `4` the following is possible:

Simplest Instantiation of `Contract`:

```typescript
const contract = new Contract(abi)
```

- `contract` has it's own independent `Web3Config` copied from the newly exported `Web3ConfigOptionsDefaults` export (`2.a.1`)
- Default values are only updated by directly setting on instance e.g. `contract.defaultBlock = 'pending'`

---

Instantiation with Global `Web3Context`:

```typescript
const web3Context = new Web3Context();
const contract = new Contract(abi, web3Context);
```

- `contract.config` is a reference to `web3Context.config`
- Any changes to `web3Context`'s properties are automatically reflected in `contract`
- Directly setting values on the instance overrides `web3Context`'s settings, but does not change `web3Context`'s values
    - This decouples the specific property from `web3Context`, all other properties are still references
    - Decoupled properties can be re-coupled via `Web3Context.link`

```typescript
const web3Context = new Web3Context();
web3Context.defaultBlock; // latest

const contract = new Contract(abi, web3Context);
contract.defaultBlock // latest

web3Context.defaultBlock = 'pending';
web3Context.defaultBlock; // pending
contract.defaultBlock // pending

contract.defaultBlock = 'earliest';
web3Context.defaultBlock; // pending
contract.defaultBlock // earliest

web3Context.defaultBlock = 'latest';
web3Context.defaultBlock; // latest
// contract.defaultBlock is now decoupled from web3Context updates, but all other properties are still linked
contract.defaultBlock // earliest
```

---

Linking the Configs of Several Instances of `Contract`

```typescript
const globalWeb3Context = new Web3Context('myprovider');
const configuredWeb3Context = new Web3Context({
    provider: globalWeb3Context.provider,
    config: { defaultBlock: 'earliest' }
});

const contract = new Contract(abi, configuredWeb3Context);
const contract2 = new Contract(abi, configuredWeb3Context);
const contract3 = new Contract(abi, configuredWeb3Context);

const contractWithGlobalContext = new Contract(abi, globalWeb3Context);

contract.defaultBlock // earliest
contract2.defaultBlock // earliest
contract3.defaultBlock // earliest

contractWithGlobalContext.defaultBlock // latest

configuredWeb3Context.defaultBlock = 'pending';

contract.defaultBlock // pending
contract2.defaultBlock // pending
contract3.defaultBlock // pending

contractWithGlobalContext.defaultBlock // latest

globalWeb3Context.defaultBlock = 'earliest'

contract.defaultBlock // pending
contract2.defaultBlock // pending
contract3.defaultBlock // pending

contractWithGlobalContext.defaultBlock // earliest
```

I don't think this is the most user friendly way to do this, but it means we don't have to make any major code changes and reinforces the idea of `Web3Context` being a grouping of desired web3 config details. This method would replace the use of `static` properties in `Contract`, so if a user wants several instances of `Contract` to share config details then they simply create a `Web3Context` configured as desired and pass it to the instances. Because instances of `Contract` will store references to the given `Web3Context`, changes will be updated automatically and the decoupling feature mentioned in `Instantiation with Global Web3Context` section still applies for each instance

Reply:
sounds good to me!  thanks for the detailed comment @spacesailor24  !
let's hear what the others think (@jdevcs @Muhammad-Altabba ).

Reply:
Thanks Wyatt and Muhammad .

This PR should add minimum functionality around defaults, so:
- remove static layer of defaults 

and only have:

1. global defaults
2. contract instance defaults

For all new features discussed in this thread like:
- reset defaults function ( by using defaults .ts file instead of .env file )
- decoupling global default with contract instance

@nikoulai please create new issue and add **feature request** label for discussion and adding in 4.x after RC.0.